cpp/src/barretenberg/dsl/acir_format/acir_to_constraint_buf.hpp

```diff
@@ -0,0 +1,279 @@
+#pragma once
+#include "acir_format.hpp"
+#include "barretenberg/common/container.hpp"
+#include "barretenberg/dsl/acir_format/blake2s_constraint.hpp"
+#include "barretenberg/dsl/acir_format/block_constraint.hpp"
+#include "barretenberg/dsl/acir_format/ecdsa_secp256k1.hpp"
+#include "barretenberg/dsl/acir_format/hash_to_field.hpp"
+#include "barretenberg/dsl/acir_format/keccak_constraint.hpp"
+#include "barretenberg/dsl/acir_format/logic_constraint.hpp"
+#include "barretenberg/dsl/acir_format/pedersen.hpp"
+#include "barretenberg/dsl/acir_format/range_constraint.hpp"
+#include "barretenberg/dsl/acir_format/recursion_constraint.hpp"
+#include "barretenberg/dsl/acir_format/schnorr_verify.hpp"
+#include "barretenberg/dsl/acir_format/sha256_constraint.hpp"
+#include "barretenberg/proof_system/arithmetization/gate_data.hpp"
+#include "serde/index.hpp"
+#include <iterator>
+
+namespace acir_format {
+
+poly_triple serialize_arithmetic_gate(Circuit::Expression const& arg)
+{
+    poly_triple pt{
+        .a = 0,
+        .b = 0,
+        .c = 0,
+        .q_m = 0,
+        .q_l = 0,
+        .q_r = 0,
+        .q_o = 0,
+        .q_c = 0,
+    };
+    // Think this never longer than 1?
+    for (const auto& e : arg.mul_terms) {
+        uint256_t qm(std::get<0>(e));
+        uint32_t a = std::get<1>(e).value;
+        uint32_t b = std::get<2>(e).value;
+        pt.q_m = qm;
+        pt.a = a;
+        pt.b = b;
+    }
+    for (const auto& e : arg.linear_combinations) {
+        barretenberg::fr x(uint256_t(std::get<0>(e)));
+        uint32_t witness = std::get<1>(e).value;
+
+        if (pt.a == 0 || pt.a == witness) {
+            pt.a = witness;
+            pt.q_l = x;
+        } else if (pt.b == 0 || pt.b == witness) {
+            pt.b = witness;
+            pt.q_r = x;
+        } else if (pt.c == 0 || pt.c == witness) {
+            pt.c = witness;
+            pt.q_o = x;
+        } else {
+            throw_or_abort("Cannot assign linear term to a constrain of width 3");
+        }
+    }
+    pt.q_c = uint256_t(arg.q_c);
+    return pt;
+}
+
+void handle_arithmetic(Circuit::Opcode::Arithmetic const& arg, acir_format& af)
+{
+    af.constraints.push_back(serialize_arithmetic_gate(arg.value));
+}
+
+void handle_blackbox_func_call(Circuit::Opcode::BlackBoxFuncCall const& arg, acir_format& af)
+{
+    std::visit(
+        [&](auto&& arg) {
+            using T = std::decay_t<decltype(arg)>;
```

Comment:
Thoughts on this instead being an overloaded function call? Feels a bit clever :)